### ISSUE TITLE
code problems

### ISSUE THREAD

**Issue by cheng-px** [azinmatin/prince#1]
How does the Amazon data use in the PRINCE？Can you show some code examples?

**Comment by wayalhruhi** [comment]
how to we use it with the dataset?

**Comment by wayalhruhi** [comment]
> How does the Amazon data use in the PRINCE？Can you show some code examples?

did you figure it out?

**Comment by azinmatin** [comment]
I used the recent Amazon customer review dataset : https://s3.amazonaws.com/amazon-reviews-pds/readme.html
Each line contains information on a user and an item she rated and reviewed. So, you have to create a graph, add a node for each user and item, and for each line in the dataset insert an edge (labeled as rated or reviewed) between the corresponding user and item nodes. You can add weights to the edges as well. The graph I generated was weighted (check out [RecWalk paper](https://www.nikolako.net/papers/ACM_WSDM2019_RecWalk.pdf) to see how these weights are computed) and contained category nodes, too. These nodes connect the items of the same category together. Using this graph, the rest is straightforward. You need to initialize a CFE instance:
CFE(graph, p, r, alpha, epsilon)

and find the counterfactual explanations using the following function: 
cfe_instance.cfe_item_centric_algo_poly(source_node, explained_node, candidate_nodes)

**Comment by cheng-px** [comment]
> > How does the Amazon data use in the PRINCE？Can you show some code examples?
> 
> did you figure it out?

thx，I got it.

**Comment by wayalhruhi** [comment]
thanks, it's working.

**Comment by wayalhruhi** [comment]
Could you explain what p and r are in CFE(graph, p, r, alpha, epsilon)?
Thanks.

**Comment by azinmatin** [comment]
Sure. P is a dictionary object whose keys are items and for each item, the value is PPR(., item), which is again a dictionary {item2: PPR(item2, item)}. The PageRank scores are computed using local_reverse_push to guarantee epsilon-approximation. Variable r is similar to p, except that it captures the residual value for each PageRank score.

**Comment by wayalhruhi** [comment]
Thanks, but local_reverse_push, only gives PPR for connected nodes right? What if I need PPR for all items, wrt to a certain item2.

**Comment by azinmatin** [comment]
Yes, the graph must be connected. So, reverse local push for item i gives
you the PPR(., i). If you need PPR(i, .), you can use the PageRank function
in networkx library, but then the approximation guarantees would not be the
same.

**Comment by wayalhruhi** [comment]
Thanks a lot.
